cmake/Provenance.cmake

```diff
@@ -0,0 +1,119 @@
+endif ()
+
+if (NOT DEFINED post_configure_dir)
+    set(post_configure_dir ${CMAKE_SOURCE_DIR}/src/generated)
+endif ()
+
+set(pre_configure_file ${pre_configure_dir}/provenance.cpp.in)
+set(post_configure_file ${post_configure_dir}/provenance.cpp)
+
+function(CheckGitWrite git_hash)
+    file(WRITE ${CMAKE_BINARY_DIR}/git-state.txt ${git_hash})
+endfunction()
+
+function(CheckGitRead git_hash)
+    if (EXISTS ${CMAKE_BINARY_DIR}/git-state.txt)
+        file(STRINGS ${CMAKE_BINARY_DIR}/git-state.txt CONTENT)
+        LIST(GET CONTENT 0 var)
+
+        set(${git_hash} ${var} PARENT_SCOPE)
+    endif ()
+endfunction()
+
+function(CheckGitVersion)
+    # Get the latest abbreviated commit hash of the working branch
+    execute_process(
+        COMMAND git log -1 --format=%h
+        WORKING_DIRECTORY ${CMAKE_CURRENT_LIST_DIR}
+        OUTPUT_VARIABLE GIT_HASH
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+        )
+
+    # Get the git branch
+    execute_process(
+        COMMAND git rev-parse --abbrev-ref HEAD
+        WORKING_DIRECTORY ${CMAKE_CURRENT_LIST_DIR}
+        OUTPUT_VARIABLE GIT_BRANCH
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+    )
+
+    CheckGitRead(GIT_HASH_CACHE)
+    if (NOT EXISTS ${post_configure_dir})
+        file(MAKE_DIRECTORY ${post_configure_dir})
+    endif ()
+
+    if (NOT EXISTS ${post_configure_dir}/provenance.hpp)
+        file(COPY ${pre_configure_dir}/provenance.hpp DESTINATION ${post_configure_dir})
+    endif()
+
+    if (NOT DEFINED GIT_HASH_CACHE)
+        set(GIT_HASH_CACHE "INVALID")
+    endif ()
+
+    # Only update the provenance.cpp if the hash has changed. This will
+    # prevent us from rebuilding the project more than we need to.
+    if (NOT ${GIT_HASH} STREQUAL ${GIT_HASH_CACHE} OR NOT EXISTS ${post_configure_file})
+        # Set che GIT_HASH_CACHE variable the next build won't have
+        # to regenerate the source file.
+        CheckGitWrite(${GIT_HASH})
+
+        configure_file(${pre_configure_file} ${post_configure_file} @ONLY)
+    endif ()
+
+endfunction()
+
+function(CheckGitSetup)
+
+    add_custom_target(ParthenonAlwaysCheckGit COMMAND ${CMAKE_COMMAND}
+        -DRUN_CHECK_GIT_VERSION=1
+        -Dpre_configure_dir=${pre_configure_dir}
+        -Dpost_configure_file=${post_configure_dir}
+        -DGIT_HASH_CACHE=${GIT_HASH_CACHE}
+        -P ${CURRENT_LIST_DIR}/Provenance.cmake
+        BYPRODUCTS ${post_configure_file}
+        )
+
+        add_library(git_version ${CMAKE_SOURCE_DIR}/src/generated/provenance.cpp)
+    target_include_directories(git_version PUBLIC ${CMAKE_BINARY_DIR}/generated)
+    add_dependencies(git_version ParthenonAlwaysCheckGit)
+
+    CheckGitVersion()
+endfunction()
+
+# This is used to run this function from an external cmake process.
+if (RUN_CHECK_GIT_VERSION)
+    CheckGitVersion()
+endif ()
+
+# Other information:
+# Compiler
+# build options
+# timestamp
+string(TIMESTAMP PARTH_BUILD_TIMESTAMP "%Y-%m-%d %H:%M:%S %Z")
```

Comment:
Does this give the timestamp when `cmake ..` is called or when `cmake --build . ` is called? I think the former, but I guess enforcing the latter would mean forcing a rebuild of `provenance.cpp` every time you call `cmake --build .`, as well as relinking everything.